---
title: "RDS"
author: "YueanZhao"
date: '2023-05-13'
output: html_document
---
# Integration
## I import data

```{r}
HB.genes <- c("HBA1","HBA2","HBB","HBD","HBE1","HBG1","HBG2","HBM","HBQ1","HBZ")
sample_list <- list()
for(i in order_origID){
  sample_list[[i]] <- Read10X(paste0("data_transcriptome/", i,"/filtered_feature_bc_matrix/"))
  sample_list[[i]] <- CreateSeuratObject(sample_list[[i]], project = i, min.cells = 3, min.features = 200)
  sample_list[[i]] <- RenameCells(sample_list[[i]], add.cell.id = i)   
  sample_list[[i]][["percent.mt"]] <- PercentageFeatureSet(sample_list[[i]], pattern = "^MT-")
  sample_list[[i]][["percent.rb"]] <- PercentageFeatureSet(sample_list[[i]], pattern = "^RP[LS]")
  sample_list[[i]][["percent.HB"]]<-PercentageFeatureSet(sample_list[[i]], features= CaseMatch(HB.genes, rownames(sample_list[[i]]))) 
}
```

## II integration-rpca
```{r}
for (m in 1:length(sample_list)) {
  sample_list[[m]] <- subset(sample_list[[m]], subset = nFeature_RNA > 200  & percent.mt < 10)
}

for (m in 1:length(sample_list)) {
  sample_list[[m]] <- NormalizeData(sample_list[[m]], 
                                    normalization.method = "LogNormalize", scale.factor = 10000) %>% 
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000)
}

features <- SelectIntegrationFeatures(object.list = sample_list)
sample_list <- lapply(X = sample_list, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})

cell.anchors <- FindIntegrationAnchors(object.list = sample_list, 
                                       reference = c(5, 1, 20, 9, 18, 26, 34, 19, 33, 36, 39, 40),
                                       reduction = "rpca", 
                                       dims = 1:50)
scRNA <- IntegrateData(anchorset = cell.anchors, dims = 1:50,k.weight=80)
```

```{r}
DefaultAssay(scRNA) <- "integrated"
scRNA <- ScaleData(scRNA,vars.to.regress=c("percent.mt","percent.HB","nFeature_RNA"))
scRNA <- RunPCA(scRNA,verbose = FALSE)
scRNA <- FindNeighbors(scRNA, dims = 1:30, k.param = 20)
scRNA <- FindClusters(scRNA, resolution = 0.5)
scRNA <- RunUMAP(scRNA, dims = 1:30)
```


```{r}
# rm low quality cluster
scRNA = scRNA[,scRNA@meta.data$seurat_clusters != 6]
scRNA = scRNA[,scRNA@meta.data$seurat_clusters != 12]

scRNA <- RunUMAP(scRNA, dims = 1:30)
```

## III cell type annotation
```{r}
new.cluster.ids <- c(`0` = "CD8+ T cells", 
                     `1` = "CD4+ T cells",
                     `2` = "Macrophages", 
                     `3` = "NK cells", 
                     `4` = "Neutrophils",
                     `5` = "CD4+ T cells",
                     #`6` = "low quality",
                     `7` = "CD14+ Monocytes", 
                     `8` = "Apoptotic T cells", 
                     `9` = "Megakaryocytes",
                     `10` = "B cells",
                     `11` = "B cells",
                     #`12` = "low quality",
                     `13` = "Epithelial cells",
                     `14` = "Cycling",
                     `15` = "Plasma cells",
                     `16` = "CD8+ T cells",
                     `17` = "Epithelial cells", 
                     `18` = "CD8+ T cells" ,
                     `19` = "CD8+ T cells",
                     `20` = "CD16+ Monocytes",
                     `21` = "CD4+ T cells",
                     `22` = "CD8+ T cells",
                     `23` = "cDCs" ,
                     `24` = "B cells",
                     `25` = "Fibroblasts",
                     `26` = "Mast cells",
                     `27` = "pDCs",
                     `28` = "Endothelial cells",
                     `29` = "CD8+ T cells",
                     `30` = "Macrophages"
)

names(new.cluster.ids) <- levels(scRNA)
scRNA <- RenameIdents(scRNA, new.cluster.ids)
scRNA[["cell_type"]] <- scRNA@active.ident
DefaultAssay(scRNA) <- "RNA"
```

```{r}
scRNA[["cell_type_main"]] <- NA
scRNA$cell_type_main[scRNA$cell_type %in% c("NK cells") ] <- "NK cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("CD8+ T cells") ] <- "T cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("CD4+ T cells") ] <- "T cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Apoptotic T cells") ] <- "T cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Cycling") ] <- "Cycling"
scRNA$cell_type_main[scRNA$cell_type %in% c("B cells") ] <- "B cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Plasma cells") ] <- "Plasma cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Megakaryocytes") ] <- "Megakaryocytes"
scRNA$cell_type_main[scRNA$cell_type %in% c("Neutrophils") ] <- "Myeloid cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("CD14+ Monocytes") ] <- "Myeloid cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Macrophages") ] <- "Myeloid cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("CD16+ Monocytes") ] <- "Myeloid cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("pDCs") ] <- "DCs"
scRNA$cell_type_main[scRNA$cell_type %in% c("cDCs") ] <- "DCs"
scRNA$cell_type_main[scRNA$cell_type %in% c("Mast cells") ] <- "Mast cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Fibroblasts") ] <- "Parenchymal cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Epithelial cells") ] <- "Parenchymal cells"
scRNA$cell_type_main[scRNA$cell_type %in% c("Endothelial cells") ] <- "Parenchymal cells"
```


## IV add meta.data: patient/sample_type/disease/patientID/orig.ident_ID
```{r}
#patient
scRNA[["patient"]] <- (str_split_fixed(scRNA$orig.ident,"_",2))[,1]
#sample_type
scRNA[["sample_type"]] <- (str_split_fixed(scRNA$orig.ident,"_",2))[,2]
#disease
scRNA[["disease"]] <- NA
scRNA$disease[scRNA$patient == "P1"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P2"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P3"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P4"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P5"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P6"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P7"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P8"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P9"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P10"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P11"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "P12"] <- "Pneumonia"
scRNA$disease[scRNA$patient == "C1"] <- "Ctrl"
scRNA$disease[scRNA$patient == "C2"] <- "Ctrl"
scRNA$disease[scRNA$patient == "C3"] <- "Ctrl"
scRNA$disease[scRNA$patient == "C4"] <- "Ctrl"
scRNA$disease[scRNA$patient == "C5"] <- "Ctrl"

#bacterial
scRNA[["bacterial"]] <- NA
scRNA$bacterial[scRNA$patient == "P1"] <- "bacterial"
scRNA$bacterial[scRNA$patient == "P2"] <- "bacterial"
scRNA$bacterial[scRNA$patient == "P3"] <- "bacterial"
scRNA$bacterial[scRNA$patient == "P4"] <- "bacterial"
scRNA$bacterial[scRNA$patient == "P5"] <- "bacterial"
scRNA$bacterial[scRNA$patient == "P6"] <- "non-bacterial"
scRNA$bacterial[scRNA$patient == "P7"] <- "non-bacterial"
scRNA$bacterial[scRNA$patient == "P8"] <- "non-bacterial"
scRNA$bacterial[scRNA$patient == "P9"] <- "non-bacterial"
scRNA$bacterial[scRNA$patient == "P10"] <- "non-bacterial"
scRNA$bacterial[scRNA$patient == "P11"] <- "non-bacterial"
scRNA$bacterial[scRNA$patient == "P12"] <- "non-bacterial"
scRNA$bacterial[scRNA$patient == "C1"] <- "ctrl"
scRNA$bacterial[scRNA$patient == "C2"] <- "ctrl"
scRNA$bacterial[scRNA$patient == "C3"] <- "ctrl"
scRNA$bacterial[scRNA$patient == "C4"] <- "ctrl"
scRNA$bacterial[scRNA$patient == "C5"] <- "ctrl"
```

## saveRDS: scRNA integrated annotated
```{r}
saveRDS(scRNA,"rds/scRNA_46_anno.rds")
```

# T&NK
## I re-integration
```{r}
scRNAsubT = scRNA[,scRNA@meta.data$seurat_clusters %in% c(0,3,1,5,19,29,21,22,8,14,16,18)]

scRNAsubT[["sample_recomb"]] <- ifelse(scRNAsubT@meta.data$orig.ident %in% c("P1_BALF","P1_PBMC"), 
                                      scRNAsubT[["sample_recomb"]]<-"P1_BLEND", 
                                      (ifelse(scRNAsubT@meta.data$orig.ident %in% c("P9_BALF","P9_PBMC"), 
                                              scRNAsubT[["sample_recomb"]]<-"P9_BLEND", 
                                              (ifelse(scRNAsubT@meta.data$orig.ident %in% c("P10_PBMC","P10_BALF"), 
                                                      scRNAsubT[["sample_recomb"]]<-"P10_BLEND", 
                                                      scRNAsubT[["sample_recomb"]]<- scRNAsubT@meta.data$orig.ident)))))


DefaultAssay(scRNAsubT)<- "RNA"
seurat_list <- SplitObject(scRNAsubT, split.by = "sample_recomb")

seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- NormalizeData(x, verbose = FALSE)
  x <- FindVariableFeatures(x, verbose = FALSE ,nfeatures = 2000)
})

features <- SelectIntegrationFeatures(object.list = seurat_list)
seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})

anchors <- FindIntegrationAnchors(object.list = seurat_list,reduction = "rpca",dims = 1:50)
scRNAsubT <- IntegrateData(anchorset = anchors, dims = 1:50)

scRNAsubT <- ScaleData(scRNAsubT, vars.to.regress = "percent.rb")
scRNAsubT <- RunPCA(scRNAsubT, verbose = FALSE)
ElbowPlot(scRNAsubT, ndims=50, reduction="pca")
scRNAsubT <- FindNeighbors(scRNAsubT, dims = 1:15, k.param = 20)
scRNAsubT <- FindClusters(scRNAsubT, resolution = 0.7)
scRNAsubT <- RunUMAP(scRNAsubT, dims = 1:15)
```

## saveRDS: scRNAsubT w/ apoptotic w/o annotation
```{r}
saveRDS(scRNAsubT,"rds/scRNAsubT&NK_all_apoptotic.rds")
```

## II re-RunUMAP w/o apoptotic doublet
```{r}
scRNAsubT <- readRDS("rds/scRNAsubT&NK_all_apoptotic.rds")
```

```{r}
scRNAsubT = scRNAsubT[,!(scRNAsubT@meta.data$seurat_clusters %in% c(13,21,22,16,9,20,23))]
scRNAsubT <- RunUMAP(scRNAsubT, dims = 1:15)
```

## III cell type annotation
```{r}
new.cluster.ids <- c(`0` = "CD8_GP6H_emra", 
                     `1` = "CD4_CCR7_naive",
                     `2` = "NK_TYROBP", 
                     `3` = "CD4_GP6K_m", 
                     `4` = "CD8_ZNF683_rm",
                     `5` = "NKT_KLRC2",
                     `6` = "CD4_CCR6_Th",
                     `7` = "CD8_GP6K_em", 
                     `8` = "CD4_RBPJ_Th", 
                     `10` = "CD4_GNLY_CTL",
                     `11` = "CD8_TYROBP_NK-like",
                     `12` = "Cycling",
                     `14` = "GDT",
                     `15` = "CD8_CCR7_naïve",
                     `17` = "Treg",
                     `18` = "MAIT",
                     `19` = "NK_GP6K"
)

names(new.cluster.ids) <- levels(scRNAsubT)
scRNAsubT <- RenameIdents(scRNAsubT, new.cluster.ids)
scRNAsubT[["cell_type"]] <- scRNAsubT@active.ident
```

```{r}
scRNAsubT[["cell_type_main"]] <- NA
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD4_CCR7_naive") ] <- "CD4_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD4_GP6K_m") ] <- "CD4_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD4_CCR6_Th") ] <- "CD4_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD4_RBPJ_Th") ] <- "CD4_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD4_GNLY_CTL") ] <- "CD4_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("Treg") ] <- "CD4_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD8_CCR7_naïve") ] <- "CD8_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD8_ZNF683_rm") ] <- "CD8_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD8_GP6K_em") ] <- "CD8_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD8_GP6H_emra") ] <- "CD8_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("CD8_TYROBP_NK-like") ] <- "CD8_T"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("NKT_KLRC2") ] <- "NK/NKT"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("NK_TYROBP") ] <- "NK/NKT"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("NK_GP6K") ] <- "NK/NKT"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("GDT") ] <- "GDT"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("MAIT") ] <- "MAIT"
scRNAsubT$cell_type_main[scRNAsubT$cell_type %in% c("Cycling") ] <- "Cycling"
```

```{r}
T_order <- c("CD4_CCR7_naive","CD4_GP6K_m","CD4_CCR6_Th","CD4_RBPJ_Th","CD4_GNLY_CTL","Treg",
             "CD8_CCR7_naïve","CD8_ZNF683_rm","CD8_GP6K_em","CD8_GP6H_emra","CD8_TYROBP_NK-like",
             "NKT_KLRC2","NK_TYROBP","NK_GP6K","GDT","MAIT","Cycling")
scRNAsubT@active.ident<- factor(scRNAsubT@active.ident, levels =T_order)
```

## IV combine TCR and scRNAsubT
### import TCR data
```{r}
contig_list <- list()
for(i in order_origID[4:46]){
  contig_list[[i]] <- read.csv(paste0("data_tcr/", i,"/filtered_contig_annotations.csv"))
}
```

### combine contigs w/ scRNAsubT (CTaa)

```{r}
combined <- combineTCR(contig_list, 
                       samples = c("P2","P2","P2","P3","P3","P3","P4","P4","P4",
                                   "P5","P5","P5","P6","P6","P6","P7","P7","P7",
                                   "P8","P8","P8","P9","P9","P9","P10","P10","P10",
                                   "P11","P11","P11","P12","P12","P12",
                                   "C1","C1","C2","C2","C3","C3",
                                   "C4","C4","C5","C5"), 
                       ID = c("PBMC","Lung","BALF","PBMC","Lung","BALF","PBMC","Lung","BALF",
                              "PBMC","Lung","BALF","PBMC","Lung","BALF","PBMC","Lung","BALF",
                              "PBMC","Lung","BALF","PBMC","Lung","BALF","PBMC","Lung","BALF",
                              "PBMC","Lung","BALF","PBMC","Lung","BALF",
                              "PBMC","Lung","PBMC","Lung","PBMC","Lung",
                              "PBMC","Lung","PBMC","Lung"),
                       cells ="T-AB")
```

```{r}
scRNAsubT <- combineExpression(combined, scRNAsubT, 
                           cloneCall="aa", 
                           groupBy = "sample",
                           proportion = FALSE, 
                           cloneTypes=c(Single=1, Small=5, Medium=20, Large=100, Hyperexpanded=2000))
```

## V define TCR_ID based on beta chain frequency in each patient
### TCR_type: AB, beta only, aC5ha only, doublet, NA
```{r}
scRNAsubT[["TCR_type"]] <- NA
scRNAsubT[["TCR_type"]] <- ifelse(scRNAsubT@meta.data$CTaa %in% "<NA>", scRNAsubT[["TCR_type"]]<-"NA", 
                                   (ifelse(stringr::str_detect(scRNAsubT@meta.data$CTaa, ';'), 
                                           scRNAsubT[["TCR_type"]]<-"doublet", 
                                           (ifelse(stringr::str_detect(scRNAsubT@meta.data$CTaa, '_NA'), 
                                                   scRNAsubT[["TCR_type"]]<-"alpha only", 
                                                   (ifelse(stringr::str_detect(scRNAsubT@meta.data$CTaa, 'NA_'), 
                                                           scRNAsubT[["TCR_type"]]<-"beta only", 
                                                           scRNAsubT[["TCR_type"]]<- "AB")))))))
```

### a chain & b chain
```{r}
scRNAsubT[["a_chain"]] <- (str_split_fixed(scRNAsubT$CTaa,"_",2))[,1]
scRNAsubT[["b_chain"]] <- (str_split_fixed(scRNAsubT$CTaa,"_",2))[,2]
```

### freq_bchain_patient & bchain_ID_patient
```{r}
ob.list <- SplitObject(scRNAsubT[,!(scRNAsubT$patient %in% c("P1"))], split.by = "patient")
names(ob.list)
```

```{r}
for(i in 1:16){
ob.list[[i]] <- ob.list[[i]][,ob.list[[i]]$TCR_type %in% c("beta only", "AB")]
aa <- as.data.frame(table(ob.list[[i]]$b_chain)) %>% dplyr::arrange(desc(Freq))
aa[["rank"]] <- order(-aa[,2])
names(aa)[1] <- "b_chain"

ob.list[[i]]$bchain_ID_patient <- aa$rank[match(ob.list[[i]]$b_chain, aa$b_chain)]
ob.list[[i]]$freq_bchain_patient <- aa$Freq[match(ob.list[[i]]$b_chain, aa$b_chain)]
}
```

```{r}
scRNAsubT[["bchain_ID_patient"]] <- NA
scRNAsubT[["freq_bchain_patient"]] <-NA

for(i in 1:16){
scRNAsubT$bchain_ID_patient[scRNAsubT$patient %in% names(ob.list)[i]] <- ob.list[[i]]$bchain_ID_patient[match(rownames(scRNAsubT[,scRNAsubT$patient %in% names(ob.list)[i]]@meta.data), rownames(ob.list[[i]]@meta.data))]

scRNAsubT$freq_bchain_patient[scRNAsubT$patient %in% names(ob.list)[i]] <- ob.list[[i]]$freq_bchain_patient[match(rownames(scRNAsubT[,scRNAsubT$patient %in% names(ob.list)[i]]@meta.data), rownames(ob.list[[i]]@meta.data))]
}
```

### TCR_ID
```{r}
scRNAsubT[["T"]] <- "T"

scRNAsubT@meta.data <- tidyr::unite(scRNAsubT@meta.data,"TCR_ID","patient","T","bchain_ID_patient",sep="_",remove=FALSE)
meta <- tidyr::unite(scRNAsubT@meta.data,"TCR_ID","patient","T","bchain_ID_patient",sep="_",remove=FALSE)

scRNAsubT[["TCR_ID"]] <- ifelse(stringr::str_detect(scRNAsubT@meta.data$TCR_ID, '_NA'), 
                               scRNAsubT[["TCR_ID"]] < -NA, 
                               scRNAsubT[["TCR_ID"]] <- meta$TCR_ID)

scRNAsubT[["T"]] <- NULL
```

### re-define TCR clonetype based on TCR_ID(freq_bchain_patient)
```{r}
scRNAsubT[["TCR_clonetype"]] <- ifelse(scRNAsubT$freq_bchain_patient == 1, scRNAsubT[["TCR_clonetype"]]<-"Single (X = 1)",
                                       (ifelse(scRNAsubT$freq_bchain_patient <= 5, scRNAsubT[["TCR_clonetype"]]<-"Small (1 < X <= 5)", 
                                               (ifelse(scRNAsubT$freq_bchain_patient <= 20, scRNAsubT[["TCR_clonetype"]]<-"Medium (5 < X <= 20)", 
                                                       (ifelse(scRNAsubT$freq_bchain_patient <= 100,
                                                               scRNAsubT[["TCR_clonetype"]]<-"Large (20 < X <= 100)", 
                                                               scRNAsubT[["TCR_clonetype"]]<-"Hyperexpanded (100 < X <= 2000)")))))))
scRNAsubT[["cloneType"]] <- NULL
```

## save RDS: scRNAsubT&NK w/o apoptotic & doublets, w/ annotation TCR_ID
```{r}
saveRDS(scRNAsubT,"rds/scRNAsubT&NK_anno_TCRID.rds")
```

# T8
## I subset T8 & re-RunPCA
```{r}
scRNAsubT8 <- scRNAsubT[,scRNAsubT@meta.data$seurat_clusters %in% c(4,7,0,11,18,15,12)]
```

```{r}
DefaultAssay(scRNAsubT8) <- "integrated"
scRNAsubT8 <- RunPCA(scRNAsubT8)
ElbowPlot(scRNAsubT8, ndims = 50)
scRNAsubT8 <- FindNeighbors(scRNAsubT8, dims = 1:15, k.param = 20)
scRNAsubT8 <- FindClusters(scRNAsubT8, resolution = 0.7)
scRNAsubT8 <- RunUMAP(scRNAsubT8, dims = 1:15)
```

## II cell type annotation
```{r}
new.cluster.ids <- c(`0` = "T8_emra/eff.1", 
                     `1` = "T8_rm",
                     `2` = "T8_em.1", 
                     `3` = "T8_NK-like", 
                     `4` = "T8_em.2",
                     `5` = "T8_naive",
                     `6` = "MAIT",
                     `7` = "T8_cm", 
                     `8` = "T8_emra/eff.2", 
                     `9` = "Cycling",
                     `10` = "Cycling",
                     `11` = "Cycling",
                     `12` = "T8_emra.1",
                     `13` = "T8_naive",
                     `14` = "T8_rm"
)

names(new.cluster.ids) <- levels(scRNAsubT8)
scRNAsubT8 <- RenameIdents(scRNAsubT8, new.cluster.ids)

scRNAsubT8[["cell_type_T8"]] <- scRNAsubT8@active.ident
T8_order <- c("T8_naive","T8_cm","T8_em.1","T8_em.2","T8_rm",
              "T8_emra/eff.1","T8_emra/eff.2","T8_NK-like","MAIT","Cycling")
scRNAsubT8$cell_type_T8 <- factor(scRNAsubT8$cell_type_T8, levels =c(T8_order))
```


```{r}
scRNAsubT8[["cell_type_main"]] <- NA
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_naive") ] <- "T8_naive"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_cm") ] <- "T8_cm"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_em.1") ] <- "T8_em"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_em.2") ] <- "T8_em"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_rm") ] <- "T8_rm"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_emra/eff.1") ] <- "T8_emra/eff"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_emra/eff.2") ] <- "T8_emra/eff"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("T8_NK-like") ] <- "T8_NK-like"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("Cycling") ] <- "Cycling"
scRNAsubT8$cell_type_main[scRNAsubT8$cell_type_T8 %in% c("MAIT") ] <- "MAIT"

T8_order_main <- c("T8_naive","T8_cm","T8_em","T8_rm","T8_emra/eff", "T8_NK-like","MAIT","Cycling")
scRNAsubT8$cell_type_main <- factor(scRNAsubT8$cell_type_main, levels =T8_order_main)
```

## III TCR percent
### rm CD4 cycling
```{r}
DefaultAssay(scRNAsubT8) <- "RNA"
cycling_CD4 <- subset(scRNAsubT8[,scRNAsubT8$cell_type_main %in% "Cycling"], subset = CD4 > 0 & CD8A == 0)
scRNAsubT8 <-  subset(scRNAsubT8, cells =setdiff(rownames(scRNAsubT8@meta.data) , rownames(cycling_CD4@meta.data)))
```

### split by orig.ident
```{r}
ob.list <- SplitObject(scRNAsubT8[,!(scRNAsubT8$patient %in% c("P1"))], split.by = "orig.ident")
```

### B_AB_sum
```{r}
for(i in 1:43){
ob.list[[i]][["B_AB_sum"]] <- NA
ob.list[[i]]$B_AB_sum <- sum(as.data.frame(table(ob.list[[i]]$freq_bchain_patient))$Freq)
}
```

```{r}
scRNAsubT8[["B_AB_sum"]] <- NA
for(i in 1:43){
scRNAsubT8$B_AB_sum[scRNAsubT8$orig.ident %in% names(ob.list)[i]] <- ob.list[[i]]$B_AB_sum[match(rownames(scRNAsubT8[,scRNAsubT8$orig.ident %in% names(ob.list)[i]]@meta.data), rownames(ob.list[[i]]@meta.data))]
}
```

### freq_bchain_sample & bchain_ID_sample
```{r}
for(i in 1:43){
  ob.list[[i]][["freq_bchain_sample"]] <- NA
  ob.list[[i]]$freq_bchain_sample <- as.data.frame(table(ob.list[[i]][,ob.list[[i]]$freq_bchain_patient >= 2]$TCR_ID))$Freq[match(ob.list[[i]]$TCR_ID, as.data.frame(table(ob.list[[i]][,ob.list[[i]]$freq_bchain_patient >= 2]$TCR_ID))$Var1)]
  }
```

```{r}
scRNAsubT8[["freq_bchain_sample"]] <- NA

for(i in 1:43){
scRNAsubT8$freq_bchain_sample[scRNAsubT8$orig.ident %in% names(ob.list)[i]] <- ob.list[[i]]$freq_bchain_sample[match(rownames(scRNAsubT8[,scRNAsubT8$orig.ident %in% names(ob.list)[i]]@meta.data), rownames(ob.list[[i]]@meta.data))]
}
```

```{r}
for(i in 1:43){
  aa = as.data.frame(table(ob.list[[i]][,ob.list[[i]]$freq_bchain_patient >= 2]$TCR_ID)) %>% dplyr::arrange(desc(Freq))
  aa[["rank"]] <- order(-aa[,2])
  names(aa)[1] <- "TCR_ID"
  
  ob.list[[i]]$freq_bchain_sample <- aa$Freq[match(ob.list[[i]]$TCR_ID, aa$TCR_ID)]
  ob.list[[i]]$bchain_ID_sample <- aa$rank[match(ob.list[[i]]$TCR_ID, aa$TCR_ID)]
  }
```

```{r}
scRNAsubT8[["freq_bchain_sample"]] <- NA
scRNAsubT8[["bchain_ID_sample"]] <- NA

for(i in 1:43){
scRNAsubT8$freq_bchain_sample[scRNAsubT8$orig.ident %in% names(ob.list)[i]] <- ob.list[[i]]$freq_bchain_sample[match(rownames(scRNAsubT8[,scRNAsubT8$orig.ident %in% names(ob.list)[i]]@meta.data), rownames(ob.list[[i]]@meta.data))]  

scRNAsubT8$bchain_ID_sample[scRNAsubT8$orig.ident %in% names(ob.list)[i]] <- ob.list[[i]]$bchain_ID_sample[match(rownames(scRNAsubT8[,scRNAsubT8$orig.ident %in% names(ob.list)[i]]@meta.data), rownames(ob.list[[i]]@meta.data))]
}
```

### TCR percent
```{r}
for(i in 1:43){
  ob.list[[i]][["TCR_percent"]] <-NA
  ob.list[[i]]$TCR_percent <- ob.list[[i]]$freq_bchain_sample / ob.list[[i]]$B_AB_sum
  }
```

```{r}
scRNAsubT8[["TCR_percent"]] <-NA
for(i in 1:43){
scRNAsubT8$TCR_percent[scRNAsubT8$orig.ident %in% names(ob.list)[i]] <- ob.list[[i]]$TCR_percent[match(rownames(scRNAsubT8[,scRNAsubT8$orig.ident %in% names(ob.list)[i]]@meta.data), rownames(ob.list[[i]]@meta.data))]
}
```

## saveRDS
```{r}
saveRDS(scRNAsubT8,"rds/scRNAsubT8_anno_TCRpercent_Cycling.NoCD4.rds")
```

# T4
## I res 0.8 re-RunUMAP
```{r}
scRNAsub <- readRDS("/mnt/sdc/project/ZYA/P1/P1_new/real_final/figures/230517/scRNAsubT&NK_all_apoptotic.rds")
```

```{r}
scRNAsub <- FindClusters(scRNAsub, resolution = 0.8)
scRNAsubT4 <- scRNAsub[,scRNAsub@meta.data$seurat_clusters %in% c(10,7,3,5,1,17,16,18)]
DefaultAssay(scRNAsubT4) <- "integrated"
scRNAsubT4 <- RunPCA(scRNAsubT4)
scRNAsubT4 <- FindNeighbors(scRNAsubT4, dims = 1:10) %>% FindClusters(resolution = 0.5) %>% RunUMAP(dims = 1:10)
DimPlot(scRNAsubT4, label = T, raster = F)
```

## II cell type annotation
```{r}
new.cluster.ids <- c(`0` = "T4_Th1", 
                     `1` = "T4_Tn.2",
                     `2` = "T4_Tm", 
                     `3` = "T4_CTL", 
                     `4` = "T4_Th17",
                     `5` = "T4_Tn.1",
                     `6` = "T4_Th2",
                     `7` = "T4_Treg", 
                     `8` = "Cycling", 
                     `9` = "T4_Tfh",
                     `10` = "Cycling",
                     `11` = "Cycling",
                     `12` = "T4_Treg"
)

names(new.cluster.ids) <- levels(scRNAsubT4)
scRNAsubT4 <- RenameIdents(scRNAsubT4, new.cluster.ids)
scRNAsubT4[["cell_type"]] <- scRNAsubT4@active.ident
```

```{r}
scRNAsubT4[["cell_type_main"]] <- NA
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Tn.1") ] <- "T4_Tn"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Tn.2") ] <- "T4_Tn"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Th1") ] <- "T4_Th1"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Tm") ] <- "T4_Tm"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_CTL") ] <- "T4_CTL"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Th17") ] <- "T4_Th17"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Th2") ] <- "T4_Th2"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Treg") ] <- "T4_Treg"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("Cycling") ] <- "Cycling"
scRNAsubT4$cell_type_main[scRNAsubT4$cell_type %in% c("T4_Tfh") ] <- "T4_Tfh"
```

## III rm CD8 cycling
```{r}
DefaultAssay(scRNAsubT4) <- "RNA"
cycling_CD8 <- subset(scRNAsubT4[,scRNAsubT4$cell_type_main %in% "Cycling"], subset = CD4 == 0 & CD8A > 0)
scRNAsubT4 <-  subset(scRNAsubT4, 
                      cells =setdiff(rownames(scRNAsubT4@meta.data) , rownames(cycling_CD8@meta.data)))
```

## IV match TCRID
```{r}
TCR_meta <- scRNAsubT@meta.data[,c(17,20:31)]

for(i in 1:length(colnames(TCR_meta))){
scRNAsubT4[[colnames(TCR_meta)[i]]] <- NA
scRNAsubT4@meta.data[,19+i] <- TCR_meta[,i][match(rownames(scRNAsubT4@meta.data),rownames(TCR_meta))]
}
```

## saveRDS
```{r}
saveRDS(scRNAsubT4,"rds/scRNAsubT4_anno_TCRID_Cycling.NoCD8A.rds")
```

# B&Plasma
## I re-integration
```{r}
scRNAsubB <- scRNA[,scRNA@meta.data$seurat_clusters %in% c(10,11,15,24)]

DefaultAssay(scRNAsubB) <- "integrated"
scRNAsubB <- RunPCA(scRNAsubB,verbose = FALSE)
scRNAsubB <- FindNeighbors(scRNAsubB, dims = 1:15, k.param = 20)
scRNAsubB <- FindClusters(scRNAsubB, resolution = 1.5)
scRNAsubB <- RunUMAP(scRNAsubB, dims = 1:15)

scRNAsubB <- scRNAsubB[,!(scRNAsubB@meta.data$seurat_clusters %in% c(8,17))]
DefaultAssay(scRNAsubB) <- "RNA"

scRNAsubB[["sample_recomb"]] <- ifelse(scRNAsubB@meta.data$patient %in% c("C4","C2"), 
                                      scRNAsubB[["sample_recomb"]]<-"C4_C2_BLEND",
                                      scRNAsubB[["sample_recomb"]]<- scRNAsubB@meta.data$patient)
seurat_list <- SplitObject(scRNAsubB, split.by = "sample_recomb")

seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- NormalizeData(x, verbose = FALSE)
  x <- FindVariableFeatures(x, verbose = FALSE ,nfeatures = 1000)
})

features <- SelectIntegrationFeatures(object.list = seurat_list)
seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})

anchors <- FindIntegrationAnchors(object.list = seurat_list,reduction = "rpca",dims = 1:30, anchor.features = 1000)

scRNAsubB <- IntegrateData(anchorset = anchors, dims = 1:30)
scRNAsubB <- ScaleData(scRNAsubB)
scRNAsubB <- RunPCA(scRNAsubB, verbose = FALSE)
scRNAsubB <- FindNeighbors(scRNAsubB, dims = 1:25, k.param = 20)
scRNAsubB <- FindClusters(scRNAsubB, resolution = 0.7)
scRNAsubB <- RunUMAP(scRNAsubB, dims = 1:25)
rm(anchors,seurat_list)
```


```{r}
DefaultAssay(scRNAsubB) <- "RNA"
scRNAsubB <- scRNAsubB[,!(scRNAsubB@meta.data$seurat_clusters %in% c(9,12,14,16))]

seurat_list <- SplitObject(scRNAsubB, split.by = "sample_recomb")
seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- NormalizeData(x, verbose = FALSE)
  x <- FindVariableFeatures(x, verbose = FALSE ,nfeatures = 1000)
})

features <- SelectIntegrationFeatures(object.list = seurat_list)
seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})

anchors <- FindIntegrationAnchors(object.list = seurat_list,reduction = "rpca",dims = 1:30, anchor.features = 1000)
scRNAsubB <- IntegrateData(anchorset = anchors, dims = 1:30)


DefaultAssay(scRNAsubB)<-"integrated"
scRNAsubB <- ScaleData(scRNAsubB, verbose = FALSE)
scRNAsubB <- RunPCA(scRNAsubB, verbose = FALSE)
scRNAsubB <- FindNeighbors(scRNAsubB, dims = 1:15, k.param = 20)
scRNAsubB <- FindClusters(scRNAsubB, resolution = 1.1)
scRNAsubB <- RunUMAP(scRNAsubB, dims = 1:15)
scRNAsubB <- scRNAsubB[,scRNAsubB$seurat_clusters != 18]
```

## II cell type annotation
```{r}
new.cluster.ids <- c(`0` = "B_naive", 
                     `1` = "B_mem.1",
                     `2` = "B_naive", 
                     `3` = "PC.1", 
                     `4` = "B_mem.1",
                     `5` = "B_mem.2",
                     `6` = "PB",
                     `7` = "B_mem.1", 
                     `8` = "B_MZ", 
                     `9` = "B_mem.1",
                     `10` = "PC.2",
                     `11` = "B_naive",
                     `12` = "B_GC",
                     `13` = "B_naive",
                     `14` = "B_pre_mem",
                     `15` = "PB",
                     `16` = "PC.3",
                     `17` = "B_naive"
)

names(new.cluster.ids) <- levels(scRNAsubB)
scRNAsubB <- RenameIdents(scRNAsubB, new.cluster.ids)
scRNAsubB[["cell_type_B"]] <- scRNAsubB@active.ident
```

```{r}
scRNAsubB[["cell_type_main"]] <- NA
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("B_naive") ] <- "B.cells"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("B_mem.1") ] <- "B.cells"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("B_mem.2") ] <- "B.cells"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("B_MZ") ] <- "B.cells"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("B_GC") ] <- "B.cells"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("B_pre_mem") ] <- "B.cells"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("PB") ] <- "PB"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("PC.1") ] <- "PC.1"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("PC.2") ] <- "PC.2"
scRNAsubB$cell_type_main[scRNAsubB$cell_type_B %in% c("PC.3") ] <- "PC.3"
```

```{r}
scRNAsubB[["cell_type"]] <- NA
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("B_naive") ] <- "B_naive"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("B_mem.1") ] <- "B_mem"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("B_mem.2") ] <- "B_mem"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("B_MZ") ] <- "B_MZ"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("B_GC") ] <- "B_GC"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("B_pre_mem") ] <- "B_pre_mem"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("PB") ] <- "PB"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("PC.1") ] <- "PC.1"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("PC.2") ] <- "PC.2"
scRNAsubB$cell_type[scRNAsubB$cell_type_B %in% c("PC.3") ] <- "PC.3"
```

## III match BCR info
### import BCR csv
```{r}
BCR_list <- list()
BCR <- data.frame()
for(i in order_patientID[2:17]){
  BCR_list[[i]] <- read.csv(paste0("data_bcr/", i,"/",i,"_BCR.csv"))
  BCR <- rbind(BCR,BCR_list[[i]]) %>% as.data.frame()
}
write.csv(BCR,"rds/BCR_data.csv",row.names = F)
```

### isotype
```{r}
scRNAsubB[["isotype"]] <- NA
scRNAsubB$isotype <- BCR$c_call[match(rownames(scRNAsubB@meta.data), BCR$barcode)]
```

### BCR_ID
```{r}
scRNAsubB[["clone_id"]] <- NA
scRNAsubB$clone_id <- BCR$clone_id[match(rownames(scRNAsubB@meta.data), BCR$barcode)]
scRNAsubB[["B"]] <- "B"
scRNAsubB@meta.data <- tidyr::unite(scRNAsubB@meta.data,"BCR_ID","patient","B","clone_id",sep="_",remove=FALSE)
meta <- tidyr::unite(scRNAsubB@meta.data,"BCR_ID","patient","B","clone_id",sep="_",remove=FALSE)

scRNAsubB[["BCR_ID"]] <- ifelse(stringr::str_detect(scRNAsubB@meta.data$BCR_ID, '_NA'), 
                               scRNAsubB[["BCR_ID"]] < -NA, 
                               scRNAsubB[["BCR_ID"]] <- meta$BCR_ID)
scRNAsubB[["B"]] <- NULL
```

### BCR_freq
```{r}
df <- as.data.frame(table(scRNAsubB$BCR_ID)) %>% dplyr::arrange(desc(Freq))
scRNAsubB[["BCR_freq"]] <- NA
scRNAsubB$BCR_freq <- df$Freq[match(scRNAsubB$BCR_ID, df$Var1)]
```

### BCR_clonetype
```{r}
scRNAsubB[["BCR_clonetype"]] <- ifelse(scRNAsubB$BCR_freq == 1, scRNAsubB[["BCR_clonetype"]]<-"Single (n = 1)",
                                       (ifelse(scRNAsubB$BCR_freq <= 5, scRNAsubB[["BCR_clonetype"]]<-"Small (1 < n <= 5)", 
                                           (ifelse(scRNAsubB$BCR_freq <= 20, scRNAsubB[["BCR_clonetype"]]<-"Medium (5 < n <= 10)", 
                                                   (ifelse(scRNAsubB$BCR_freq <= 100,
                                                           scRNAsubB[["BCR_clonetype"]]<-"Large (10 < n <= 100)", 
                                                           scRNAsubB[["BCR_clonetype"]]<-"Hyperexpanded (n > 100)")))))))

scRNAsubB$BCR_clonetype<- factor(scRNAsubB$BCR_clonetype, 
                                 levels =c("Hyperexpanded (n > 100)",
                                           "Large (10 < n <= 100)",
                                           "Medium (5 < n <= 10)",
                                           "Small (1 < n <= 5)",
                                           "Single (n = 1)" ))
```

### SHM_freq & SHM_group
```{r}
scRNAsubB[["SHM_freq"]] <- NA
scRNAsubB$SHM_freq <- BCR$mu_freq[match(rownames(scRNAsubB@meta.data), BCR$barcode)]
scRNAsubB[["SHM_group"]] <- ifelse(scRNAsubB$SHM_freq == 0, scRNAsubB[["SHM_group"]]<-"none",
                                       (ifelse(scRNAsubB$SHM_freq <= 0.1, scRNAsubB[["SHM_group"]]<-"low", 
                                               scRNAsubB[["SHM_group"]]<-"high")))
```

## saveRDS
```{r}
saveRDS(scRNAsubB,"rds/scRNAsubB_anno_BCR.rds")
```

# Monocytes & Macrophages
## I re-integration
```{r}
scRNAsubM <- scRNA[,scRNA@meta.data$seurat_clusters %in% c(2,7,20,30)]

scRNAsubM[["sample_recomb"]] <- ifelse(scRNAsubM@meta.data$orig.ident %in% c("C3_PBMC","C3_LUNG"), 
                                   scRNAsubM[["sample_recomb"]]<-"C3_BLEND", 
                                   (ifelse(scRNAsubM@meta.data$orig.ident %in% c("P9_BALF","P9_PBMC"), 
                                           scRNAsubM[["sample_recomb"]]<-"P9_BLEND",
                                           (ifelse(scRNAsubM@meta.data$orig.ident %in% c("C2_PBMC","C2_LUNG"), 
                                                   scRNAsubM[["sample_recomb"]]<-"C2_BLEND",
                                                   (ifelse(scRNAsubM@meta.data$orig.ident %in% c("P8_PBMC","P8_BALF"), 
                                                           scRNAsubM[["sample_recomb"]]<-"P8_BLEND",
                                                           (ifelse(scRNAsubM@meta.data$orig.ident %in% c("P4_LUNG","P4_BALF"), 
                                                                   scRNAsubM[["sample_recomb"]]<-"P4_BLEND",
                                                                   scRNAsubM[["sample_recomb"]]<- scRNAsubM@meta.data$orig.ident)))))))))
```

```{r}
seurat_list <- SplitObject(scRNAsubM, split.by = "sample_recomb")

seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- NormalizeData(x, verbose = FALSE)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 1000)
})

features <- SelectIntegrationFeatures(object.list = seurat_list)
seurat_list <- lapply(X = seurat_list, FUN = function(x) {
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})

anchors <- FindIntegrationAnchors(object.list = seurat_list,reduction = "rpca",dims = 1:50)
scRNAsubM <- IntegrateData(anchorset = anchors, dims = 1:50)

scRNAsubM <- ScaleData(scRNAsubM, vars.to.regress = c("percent.rb"))
scRNAsubM <- RunPCA(scRNAsubM, verbose = FALSE)
scRNAsubM <- FindNeighbors(scRNAsubM, dims = 1:30, k.param = 20)
scRNAsubM <- FindClusters(scRNAsubM, resolution = 1.5)
scRNAsubM <- RunUMAP(scRNAsubM, dims = 1:30)
```


```{r}
scRNAsubM <- scRNAsubM[,!(scRNAsubM@meta.data$seurat_clusters %in% c(12,18,21,22,23,26,28))]
DefaultAssay(scRNAsubM) <- "integrated"
scRNAsubM <- RunPCA(scRNAsubM,verbose = FALSE)
scRNAsubM <- FindNeighbors(scRNAsubM, dims = 1:30, k.param = 20)
scRNAsubM <- FindClusters(scRNAsubM, resolution = 1.5)
scRNAsubM<- RunUMAP(scRNAsubM, dims = 1:30) 
```

```{r}
scRNAsubM <- scRNAsubM[,!(scRNAsubM@meta.data$seurat_clusters %in% c(15))]
DefaultAssay(scRNAsubM) <- "integrated"
scRNAsubM <- RunPCA(scRNAsubM,verbose = FALSE)
scRNAsubM <- FindNeighbors(scRNAsubM, dims = 1:15, k.param = 20)
scRNAsubM <- FindClusters(scRNAsubM, resolution = 0.5)
scRNAsubM<- RunUMAP(scRNAsubM, dims = 1:15) 
```

```{r}
scRNAsubM <- scRNAsubM[,!(scRNAsubM@meta.data$seurat_clusters %in% c(11))]
DefaultAssay(scRNAsubM) <- "integrated"
scRNAsubM<- RunUMAP(scRNAsubM, dims = 1:15) 
```

## II cell type annotation
```{r}
DefaultAssay(scRNAsubM) <- "RNA"
scRNAsubM[["cell_type"]] <- NA
new.cluster.ids <- c(`0` = "MDM_CCL18", 
                     `1` = "transitioning MDM_CCL2",
                     `2` = "Mono_CD14", 
                     `3` = "transitioning MDM_CCL2", 
                     `4` = "Mono_CD14",
                     `5` = "Mono_CD14_CD16",
                     `6` = "transitioning MDM_FMN1",
                     `7` = "Mono_CD14", 
                     `8` = "Mono_CD16",
                     `9` = "transitioning MDM_CCL2",
                     `10` = "AM_FABP4"
)

names(new.cluster.ids) <- levels(scRNAsubM)
scRNAsubM <- RenameIdents(scRNAsubM, new.cluster.ids)
scRNAsubM[["cell_type"]] <- scRNAsubM@active.ident
```

## saveRDS
```{r}
saveRDS(scRNAsubM, "rds/scRNAsubM_macro&mono_anno.rds")
```


# Epithelial
## I re-RunPCA

```{r}
scRNAsubE <- scRNA[,scRNA$cell_type %in% c("Epithelial cells")]
```

```{r}
DefaultAssay(scRNAsubE) <- "integrated"
scRNAsubE <- RunPCA(scRNAsubE)
scRNAsubE <- FindNeighbors(scRNAsubE, dims = 1:20, k.param = 20)
scRNAsubE <- FindClusters(scRNAsubE, resolution = 0.7)
scRNAsubE <- RunUMAP(scRNAsubE, dims = 1:20)
```

## II cell type annotation
```{r}
new.cluster.ids <- c(`0` = "AT2", 
                     `1` = "Ciliated cells",
                     `2` = "EMT_Epi", 
                     `3` = "Tuft cells", 
                     `4` = "Ciliated cells",
                     `5` = "Ciliated cells",
                     `6` = "AT1",
                     `7` = "Ciliated cells", 
                     `8` = "AT2", 
                     `9` = "Ciliated cells",
                     `10` = "Club cells",
                     `11` = "Goblet cells",
                     `12` = "EMT_Epi",
                     `13` = "EMT_Epi",
                     `14` = "Mesothelial cells",
                     `15` = "EMT_Epi",
                     `16` = "AT1",
                     `17` = "EMT_Epi"
)

names(new.cluster.ids) <- levels(scRNAsubE)
scRNAsubE <- RenameIdents(scRNAsubE, new.cluster.ids)

scRNAsubE[["cell_type"]] <- scRNAsubE@active.ident
```

## saveRDS
```{r}
saveRDS(scRNAsubE,"rds/scRNAsubE_anno.rds")
```

# P6 flu infection
```{r}
P6 <- read_csv("rds/P6_fluB_reads.csv")

scRNA_P6 <- scRNA[,scRNA$patient == "P6"]
scRNA_P6$fluB <- P6$'Influenza B virus'[match(rownames(scRNA_P6@meta.data), P6$barcode)]
scRNA_P6$fluB <- as.numeric(scRNA_P6$fluB)
scRNA_P6$fluB[scRNA_P6$fluB == 0] <- NA
```

```{r}
saveRDS(scRNA_P6,"rds/scRNA_P6_fluB.rds")
```

# P6 T8 top10 TCR MHC-I score & pattern 
```{r}
disease_pattern <- readRDS("rds/T8_disease_TCRpattern.rds")
P6_top10 <- disease_pattern[,disease_pattern$patient == "P6"]
```

```{r}
cellchat <- readRDS("cellchat/cellchat_flu_P6TCR(>10)_MHCI.rds")
aa = list(cellchat@netP)
cc = aa[[1]]$prob
dd = cc[1:55,1:55,"MHC-I"]
dd = dd[,-1]
dd = dd[1,]
```

```{r}
P6_top10[["P6_MHCI"]] <- NA
data <- as.data.frame(dd)
P6_top10$P6_MHCI <- data$dd[match(P6_top10$TCR_ID, rownames(data))]
P6_top10@meta.data$P6_MHCI <- as.numeric(P6_top10@meta.data$P6_MHCI)
```

## saveRDS
```{r}
saveRDS(P6_top10,"rds/P6_top10TCR_pattern_MHCIscore.rds")
```

